multiversx_sdk_cli/multisig.py

```diff
@@ -0,0 +1,916 @@
+        should_prepare_args_for_factory: bool,
+        guardian_and_relayer_data: GuardianRelayerData,
+        native_token_amount: int = 0,
+        abi: Optional[Abi] = None,
+        arguments: Optional[list[Any]] = None,
+    ) -> Transaction:
+        args = arguments if arguments else []
+        if should_prepare_args_for_factory:
+            args = self._convert_args_to_typed_values(args)
+
+        tx = self._factory.create_transaction_for_propose_contract_deploy_from_source(
+            sender=owner.address,
+            contract=contract,
+            gas_limit=gas_limit,
+            contract_to_copy=contract_to_copy,
+            is_upgradeable=upgradeable,
+            is_readable=readable,
+            is_payable=payable,
+            is_payable_by_sc=payable_by_sc,
+            native_token_amount=native_token_amount,
+            abi=abi,
+            arguments=args,
+        )
+        tx.nonce = nonce
+        tx.version = version
+        tx.options = options
+        tx.gas_price = gas_price
+        tx.guardian = guardian_and_relayer_data.guardian_address
+        tx.relayer = guardian_and_relayer_data.relayer_address
+
+        self.sign_transaction(
+            transaction=tx,
+            sender=owner,
+            guardian=guardian_and_relayer_data.guardian,
+            relayer=guardian_and_relayer_data.relayer,
+            guardian_service_url=guardian_and_relayer_data.guardian_service_url,
+            guardian_2fa_code=guardian_and_relayer_data.guardian_2fa_code,
+        )
+
+        return tx
+
+    def prepare_contract_upgrade_from_source_transaction(
+        self,
+        owner: IAccount,
+        nonce: int,
+        contract: Address,
+        contract_to_upgrade: Address,
+        contract_to_copy: Address,
+        gas_limit: int,
+        gas_price: int,
+        version: int,
+        options: int,
+        upgradeable: bool,
+        readable: bool,
+        payable: bool,
+        payable_by_sc: bool,
+        should_prepare_args_for_factory: bool,
+        guardian_and_relayer_data: GuardianRelayerData,
+        native_token_amount: int = 0,
+        abi: Optional[Abi] = None,
+        arguments: Optional[list[Any]] = None,
+    ) -> Transaction:
+        args = arguments if arguments else []
+        if should_prepare_args_for_factory:
+            args = self._convert_args_to_typed_values(args)
+
+        tx = self._factory.create_transaction_for_propose_contract_upgrade_from_source(
+            sender=owner.address,
+            contract=contract,
+            contract_to_upgrade=contract_to_upgrade,
+            contract_to_copy=contract_to_copy,
+            gas_limit=gas_limit,
+            is_upgradeable=upgradeable,
+            is_readable=readable,
+            is_payable=payable,
+            is_payable_by_sc=payable_by_sc,
+            native_token_amount=native_token_amount,
+            abi=abi,
+            arguments=args,
+        )
+        tx.nonce = nonce
```

Comment:
just a suggestion, I would extract this in a setMandatoryFields or something similar
``` tx.version = version
        tx.options = options
        tx.gas_price = gas_price
        tx.guardian = guardian_and_relayer_data.guardian_address
        tx.relayer = guardian_and_relayer_data.relayer_address
```